File: 3425.longest-special-path.java
#
# @lc app=leetcode id=3425 lang=java
#
# [3425] Longest Special Path
#

# @lc code=start
class Solution {
    public int[] longestSpecialPath(int[][] edges, int[] nums) {
        int n = nums.length;
        List<int[]>[] adj = new List[n];
        for (int i = 0; i < n; ++i) adj[i] = new ArrayList<>();
        for (int[] e : edges) {
            adj[e[0]].add(new int[]{e[1], e[2]});
            adj[e[1]].add(new int[]{e[0], e[2]});
        }
        int[] res = new int[]{0, Integer.MAX_VALUE}; // [max_length, min_nodes]
        Set<Integer> path = new HashSet<>();
        dfs(0, -1, adj, nums, path, 0, 1, res);
        res[1] = Math.max(1, res[1]); // at least one node
        return res;
    }
    private void dfs(int u, int parent, List<int[]>[] adj, int[] nums, Set<Integer> path, int curLen, int curNodes, int[] res) {
        if (path.contains(nums[u])) {
            if (curLen > res[0]) {
                res[0] = curLen;
                res[1] = curNodes - 1;
            } else if (curLen == res[0]) {
                res[1] = Math.min(res[1], curNodes - 1);
            }
            return;
        }
        path.add(nums[u]);
        boolean isLeaf = true;
        for (int[] nei : adj[u]) {
            if (nei[0] != parent) {
                isLeaf = false;
                dfs(nei[0], u, adj, nums, path, curLen + nei[1], curNodes + 1, res);
            }
        }
        if (isLeaf) {
            if (curLen > res[0]) {
                res[0] = curLen;
                res[1] = curNodes;
            } else if (curLen == res[0]) {
                res[1] = Math.min(res[1], curNodes);
            }
        }
        path.remove(nums[u]);
    }
}
# @lc code=end

File: 3425.longest-special-path.go
#
# @lc app=leetcode id=3425 lang=golang
#
# [3425] Longest Special Path
#

# @lc code=start
func longestSpecialPath(edges [][]int, nums []int) []int {
    n := len(nums)
    adj := make([][][2]int, n)
    for _, e := range edges {
        u, v, w := e[0], e[1], e[2]
        adj[u] = append(adj[u], [2]int{v, w})
        adj[v] = append(adj[v], [2]int{u, w})
    }
    maxLen := 0
    minNodes := n+1
    var dfs func(u, parent int, seen map[int]struct{}, curLen, curNodes int)
    dfs = func(u, parent int, seen map[int]struct{}, curLen, curNodes int) {
        if _, exists := seen[nums[u]]; exists {
            if curLen > maxLen {
                maxLen = curLen
                minNodes = curNodes
            } else if curLen == maxLen && curNodes < minNodes {
                minNodes = curNodes
            }
            return
        }
        seen[nums[u]] = struct{}{}
        extended := false
        for _, pair := range adj[u] {
            v, w := pair[0], pair[1]
            if v == parent { continue }
            extended = true
            dfs(v, u, seen, curLen + w, curNodes + 1)
        }
        if !extended {
            if curLen > maxLen {
                maxLen = curLen
                minNodes = curNodes
            } else if curLen == maxLen && curNodes < minNodes {
                minNodes = curNodes
            }
        }
        delete(seen, nums[u])
    }
    dfs(0, -1, make(map[int]struct{}), 0, 1)
    return []int{maxLen, minNodes}
}
# @lc code=end